### PR TITLE
ENH: MultiIndex.from_product infers names from inputs if not explicitly provided

### LINKED ISSUE #27292: Infer names in MultiIndex.from_product if inputs have a name attribute
It would be convenient to have `from_product` infer level names from inputs if at all possible.

### Current behavior

```python
>>> a = pd.Series([1, 2, 3], name='a')
>>> b = pd.Series(['a', 'b'], name='b')

>>> pd.MultiIndex.from_product([a, b])
MultiIndex(levels=[[1, 2, 3], ['a', 'b']],
           codes=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 0, 1]])

```

### Current workaround

```python

>>> arrs = [a, b]
>>> pd.MultiIndex.from_product(ins, names=[el.name for el in arrs])
MultiIndex(levels=[[1, 2, 3], ['a', 'b']],
           codes=[[0, 0, 1, 1, 2, 2], [0, 1, 0, 1, 0, 1]],
           names=['a', 'b'])
```

Obviously this would't make sense for lists or arrays being passed in, but in the case of a `Series`, it would be nice to have the name persisted.

---

From glancing at the source, it looks like this would be somewhat simple to implement, the naive approach might be something like:

```
        from pandas.core.arrays.categorical import _factorize_from_iterables
        from pandas.core.reshape.util import cartesian_product

        if not is_list_like(iterables):
            raise TypeError("Input must be a list / sequence of iterables.")
        elif is_iterator(iterables):
            iterables = list(iterables)

        if names is None:
            names = [el.name if hasattr(el, 'name') else None for el in iterables]

        codes, levels = _factorize_from_iterables(iterables)
        codes = cartesian_product(codes)
        return MultiIndex(levels, codes, sortorder=sortorder, names=names)
```

### PR DESCRIPTION
**Pull request by christopherzimmerman**
- [x] closes #27292
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Infers names in `MultiIndex.from_product`  and normalizes handing of `names=None` between `from_arrays` and `from_product`

### CODE REVIEW

**Review by TomAugspurger — APPROVED** [review]
One tiny comment: https://github.com/pandas-dev/pandas/pull/28417/files#r323883775 LGTM otherwise.

**Review by jreback — CHANGES_REQUESTED** [review]
tiny comment, other lgtm.

File: pandas/core/indexes/multi.py [hunk 1]
```diff
@@ -510,6 +512,11 @@ def from_product(cls, iterables, sortorder=None, names=None):
         names : list / sequence of str, optional
             Names for the levels in the index.
 
+            ..versionchanged 1.0
```
Comment (TomAugspurger) [inline pandas/core/indexes/multi.py]:
```suggestion
            .. versionchanged:: 1.0.0
```

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -510,6 +512,11 @@ def from_product(cls, iterables, sortorder=None, names=None):
...
+
+            If not explicitly provided, names will be inferred from the
```
Comment (TomAugspurger) [inline pandas/core/indexes/multi.py]:
```suggestion
               If not explicitly provided, names will be inferred from the
```

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -510,6 +512,11 @@ def from_product(cls, iterables, sortorder=None, names=None):
...
+            elements of iterables if an element has a name attribute
```
Comment (TomAugspurger) [inline pandas/core/indexes/multi.py]:
```suggestion
               elements of iterables if an element has a name attribute
```

**Review by TomAugspurger — APPROVED** [review]
Doc formatting. Looks good otherwise.

**Review by TomAugspurger — APPROVED** [review]
Verified that we do have other coverage for `from_arrays` and input arrays with names.

**Comment by TomAugspurger** [comment]
Thanks!

**Pull request merged**